Support Bison up to version 3.7

ilmari opened the pull request:
This requires copying the `YY_CAST` and `YY_ATTRIBUTE_UNUSED` macros
from the generated code, and extracting the `yysymbol_kind_t` enum if
it's defined.

We must also handle token type names with escaped double-quotes in
them, since it now names the `YYEOF` and `YYUNDEF` tokens `"end of
file"` and `"invalid token"` instead of `$end` and `$undefined`,
respectively.

ilmari committed:
Add support for Bison versions up to 3.7

This requires copying the `YY_CAST` and `YY_ATTRIBUTE_UNUSED` macros
from the generated code, and extracting the `yysymbol_kind_t` enum if
it's defined.

We must also handle token type names with escaped double-quotes in
them, since it now names the `YYEOF` and `YYUNDEF` tokens `"end of
file"` and `"invalid token"` instead of `$end` and `$undefined`,
respectively.

leonerd commented on regen_perly.pl:
```diff
@@ -76,10 +76,10 @@
 # Don't change this to add new bison versions without testing that the generated
 # files actually work :-) Win32 in particular may not like them. :-(
-unless ($version =~ /\b(1\.875[a-z]?|2\.[0134567]|3\.[0-4])\b/) { die <<EOF; }
+unless ($version =~ /\b(1\.875[a-z]?|2\.[0134567]|3\.[0-7]+)\b/) { die <<EOF; }
```
Did you mean to include this `+` at the end?

ilmari replied:
I originally made it `[0-9]+` to allow any 3.x version, but given the amount of work to get from 3.4 to 3.6, I discarded that idea as unwarrantedly optimistic, but failed to completely fix it.

Leont commented on regen_perly.pl:
```diff
@@ -90,7 +90,7 @@
 
 # creates $tmpc_file and $tmph_file
-my_system("$bison -d -o $tmpc_file $y_file");
+my_system("$bison -Wno-deprecated -d -o $tmpc_file $y_file");
```
If we do deprecated stuff, perhaps we should document what exactly?

ilmari replied:
It's the `%pure-parser` directive, which is deprecated in favor of `%define api.pure`
since Bison 2.3b (2008-05-27).

We could declare that we only support 2.4 and newer (even RHEL 6 has 2.4.1).
This only affects people hacking on the grammar, not end-users.

ilmari replied:
I've added a comment for now.

bulk88 commented:
bug from discussion in thread https://www.nntp.perl.org/group/perl.perl5.porters/2020/07/msg258091.html

ilmari committed:
Bump minimum required Bison version to 2.4

This lets us replace the deprecated `%pure-parser` directive with
`%define api.pure`, and get rid of some other conditional code.

Bison is only required for developers hacking on the grammar, since we
check in the generated code.  Bison 2.4 was released in 2008, and is
included in operating systems as old as Red Hat Enterprise Linux 6.

ilmari commented:
I've rebased this, and added another commit that bumps the minimum required version to 2.4, so we can replace the deprecated `%pure-parser` directive with `%define api.pure`.

tonycoz reviewed (approved):
This looks fine to me.

leonerd reviewed (approved):
LGTM
